sqldelight-runtime/src/main/java/com/squareup/sqldelight/SqlDelightDatabase.kt

```diff
@@ -0,0 +1,98 @@
+package com.squareup.sqldelight
+
+import android.arch.persistence.db.SupportSQLiteOpenHelper
+import android.database.sqlite.SQLiteTransactionListener
+import java.io.Closeable
+import java.util.LinkedHashSet
+import java.util.concurrent.TimeUnit
+
+/**
+ * A transaction-aware [SupportSQLiteOpenHelper] wrapper which can queue work to be performed after
+ * the active [Transaction] ends with [afterTransaction].
+ */
+abstract class SqlDelightDatabase(
+    internal val helper: SupportSQLiteOpenHelper
+) : Transactor, Closeable {
+
+  private val transactions = ThreadLocal<SqliteTransaction>()
+
+  private val transaction = object : Transaction {
+    override fun markSuccessful() {
+      helper.writableDatabase.setTransactionSuccessful()
+    }
+
+    override fun yieldIfContendedSafely(): Boolean {
+      return helper.writableDatabase.yieldIfContendedSafely()
+    }
+
+    override fun yieldIfContendedSafely(sleepAmount: Long, sleepUnit: TimeUnit): Boolean {
+      return helper.writableDatabase.yieldIfContendedSafely(sleepUnit.toMillis(sleepAmount))
+    }
+
+    override fun end() {
+      val transaction = transactions.get() ?: throw IllegalStateException("Not in transaction.")
+      val newTransaction = transaction.parent
+      transactions.set(newTransaction)
+      helper.writableDatabase.endTransaction()
+      if (transaction.commit) {
+        if (newTransaction != null) {
+          newTransaction.addAll(transaction)
```

Comment:
this is important since i changed it from sqlbrite (everything else is the same) -> actions are only run once the root transaction is complete, instead of running during nested transactions. I cant remember if we verified behavior at all but I dont think that changes to the database are committed until the root transaction is nested so it would be incorrect to notify query listeners mid-root-transaction

Reply:
SQL Brite has this behavior as well. Only the root transaction on the thread notifies.

Reply:
ah i see it now. I copied the code over differently.